### ISSUE TITLE
Infrastructure redesign brainstorming

### ISSUE THREAD

**Issue by adamrehn** [adamrehn/ue4-docker#48]
**This issue is partly a note to myself, but also an opportunity for other ue4-docker users to discuss potential design ideas and contribute their perspectives. Feedback from developers using ue4-docker as part of their own workflows is appreciated.**

A problem I've been hitting up against for some time now is that ue4-docker was originally designed for building only a small handful of container images, and the scalability of the project is somewhat hamstrung by its current architecture. The codebase gets messier and messier as new images and customisation options are added, and also whenever new Unreal Engine versions are released that require additional logic/patching or (God forbid) a new version of Visual Studio under Windows. It's reached the point where I'm hesitant to add any of the new features that I want to implement (e.g. Perforce support, cross-compilation support in Windows containers, hooks that allow custom Dockerfile instructions to be injected by the user at various points, etc.) because I can just picture how convoluted the code will look if those modifications are added to the existing design. It's time for a refactor and a new approach.

My current line of thought is that ue4-docker should split the build process into two distinct phases: an initial **generation phase** that produces a set of Dockerfiles on disk and then a **build phase** that consumes the generated Dockerfiles. This would facilitate a template-based approach whereby Dockerfiles can be customised with a high degree of flexibility, without littering the resulting code with no-ops for unused options (as is the case in the current architecture.) This could be paired with a plugin-based system for applying arbitrary transformations to the template data, allowing for an infinite number of possible configurations. I'm also toying with the idea of using a graph-based system for managing dependencies between container images, which could open up the door for user-defined images that declaratively specify their dependencies and get built alongside the predefined images in the correct order. Regardless of the exact shape that the implementation takes, my primary goal is to create a data-oriented architecture with a clean and well-abstracted API.

A few benefits that I foresee:
- Newer (or older) versions of Visual Studio can be supported by simply providing drop-in snippets for each version that can be selected by the user at generation-time.
- The addition of cross-compilation toolchains (e.g. Windows to Linux, Windows to Android) can be implemented a clean manner that doesn't impact images generated without cross-compilation support.
- If users want to implement support for new version control systems (Perforce, Mercurial, etc.), authentication mechanisms (SSH, OAuth, etc.) or even to copy the Unreal Engine source tree from the host filesystem (as in [Pull Request #47](https://github.com/adamrehn/ue4-docker/pull/47)), they can do so by simply implementing a plugin, eliminating the need to modify ue4-docker itself. Popular plugins could of course ship with ue4-docker for ease of installation.
- Injection of arbitrary user-specified Dockerfile directives can be achieved either via plugins or simply by editing the generated Dockerfiles prior to building them.
- Users who need to maintain a large set of images (e.g. game development studios who are cross-compiling for multiple platforms) can script the generation step, or even check the generated Dockerfiles into source control.

A few challenges and open questions:
- An infinite number of possible configurations makes thorough testing extremely difficult, leading to an increased likelihood of users encountering issues when building configurations that I've not tried myself.
- How can we tag images and manage inter-image dependencies in a sensible manner? Ideally all customisations should manifest themselves as suffixes in image tags so that downstream images can specify the required configuration for their base images. But with an infinite number of possible configurations, any one-to-one mapping could produce ridiculously long and unwieldy tag names.
- What's the most reasonable way to embed configuration metadata in built images so that they can be inspected after being pulled from a container registry? (Hopefully this scenario will become a reality once GitHub Package Registry comes out of beta.) Is it sensible to simply create `LABEL` directives specifying each configuration option, such that a hypothetical `ue4-docker inspect` command could simply query the labels to print a summary of the configuration that produced the image? If so, do we require all plugin authors to specify label names for their customisations and follow a reverse-DNS naming scheme to avoid collisions? How should this account for manual modifications to generated Dockerfiles, if at all?

This redesign is likely to remain an ongoing process for some time to come, and I'd appreciate feedback from other ue4-docker users as I evolve these ideas and move towards a point where I can begin implementing them.

**Comment by adamrehn** [comment]
@ImmortalEmperor, @mentalnote, @crizCraig, @TBBle, @herman-rogers

**Comment by TBBle** [comment]
A use-case I've been meaning to follow-up with here, would be to have a plugin for BuildGraph changes. In our particular use-case, we sometimes need a branch of GitHub UE4 Release tags *only* to add the BuildGraph commands for the UE4Client and UE4Server targets to the InstalledBuild.xml and InstalledBuild.filter.xml.

I'd also planned to try and work out how to make them conditional, and get those added to Epic's upstream version, but in the meantime it'd be nice to auto-apply them to any random UE4 source tree I build. That auto-applying might need per-version fixing, and I'd aim to make it generic enough to apply to all platforms, at least for the UE4Client target.

**Comment by TBBle** [comment]
On `LABEL`, presumably a plugin must have some identification so it's loaded, simply using that in the `LABEL` would work, wouldn't it? It seems reasonable to expect plugins to provide a label for images generated with that plugin, and even *strongly encourage* labels to be reverse-DNS to avoid name collisions and provide some sense of providence for the plugin, if you you have at hand is the resulting Docker image.

**Comment by adamrehn** [comment]
@TBBle that's definitely a use-case that's been on my mind too, and you're [not the only one to](https://github.com/adamrehn/ue4cli/issues/7) [run into the lack of support for those targets by default](https://github.com/adamrehn/ue4-docker/issues/49). It seems odd that there isn't already a BuildGraph flag to enable support for them, but I guess it's yet another symptom of the fact that [Epic don't consume Installed Builds internally](https://docs.unrealengine.com/en-US/Programming/Deployment/UsinganInstalledBuild/index.html#choosingthisdeploymentmethod).

**Comment by adamrehn** [comment]
With regards to `LABEL` directives, that's certainly the line of thought I've been considering. Requiring each plugin to specify a unique reverse-DNS identifier would be extremely helpful for diagnostic and debugging purposes, since ue4-docker could list the loaded plugins in its output and those lines could then be included when reporting bugs. (A maintainer email address might also be a good idea so that the appropriate plugin maintainers can be contacted when dealing with bugs arising from plugin code.)

If plugins are able to specify multiple feature flags (which might be desirable in a context where flags are related, or are mutually incompatible and the plugin needs to throw an error when a conflict is detected) then each individual flag could simply be appended to the plugin's identifier, e.g.:

```
LABEL com.organisation.plugin.flag=1
```

**Comment by dskvr** [comment]
What is the status on this refactor initiative?

**Comment by adamrehn** [comment]
@dskvr this is currently blocked by an upstream limitation regarding support for building Windows containers with BuildKit (https://github.com/moby/buildkit/issues/616) that makes it impossible to mount secrets during the build process without them leaking into the image's filesystem layers, as is possible when building Linux container images using the experimental mount types supported by BuildKit.

One of the driving factors behind the proposed two-phase design is that it will address the current incompatibilities with external build systems, which are typically designed to accept ordinary Dockerfiles (and optionally, accompanying build arguments) as their inputs, but are not designed to delegate control of the build to an arbitrary application. In the new design, ue4-docker will generate a set of Dockerfiles in filesystem directories (similar to standard base images found on Docker Hub, e.g. [Microsoft .NET](https://github.com/dotnet/dotnet-docker/tree/master/src), [Node.js](https://github.com/nodejs/docker-node/tree/master/14), etc.) and the user can store those files / check them into version control / etc. and then build them later without requiring ue4-docker at all. For this to be feasible, the [credential endpoint](https://github.com/adamrehn/ue4-docker/blob/v0.0.68/ue4docker/infrastructure/CredentialEndpoint.py) needs to be eliminated entirely, as it is currently the only component of ue4-docker that is strictly required during the build process. This is currently possible for Linux containers but not Windows containers, and providing a harness to run the credential endpoint when building Windows containers defeats the intended purpose. (Additional harnesses would also need to be created for supporting SSH credentials or Perforce credentials, and community-contributed plugins for the generation phase that use those authentication types would be forced to integrate with the harnesses under Windows, which isn't a scalable strategy.)

This is definitely still a high priority, and we're going to be needing this internally at TensorWorks as we start exploring support for targeting console platforms in on-premises installations of [Admiral](https://admiral.build) (since I'm guessing those target platforms will likely require cross-compilation from Windows rather than Linux.) There's a good chance that I'll end up continuing the work that @TBBle started on adding functional Windows container support to BuildKit, since fixing the upstream limitation seems like the only way to achieve the two-phase design with feature parity across both Windows and Linux containers.

**Comment by dskvr** [comment]
@adamrehn Thanks! What are the chances these changes will be public before Admiral is complete?

**Comment by adamrehn** [comment]
@dskvr any changes made to ue4-docker (or any of my other open source projects) will always be public as soon as they're implemented and working well enough for other users to benefit from them. The only thing I won't be able to make public are any plugins that specifically relate to console platforms, since there are legal restrictions around that, but everything else will be available for the community as soon as it's working nicely across both Windows and Linux containers.

**Comment by TBBle** [comment]
I do intend to spend more time on getting BuildKit working on Windows, but I haven't had the good stretches of time available for a personal project in a while.

Plus it kind-of got stuck on whether I needed to support containerd 1.2 as a back-end with the changes I was make, e.g. https://github.com/moby/buildkit/pull/1387, and I lost a lot of momentum on that and https://github.com/containerd/containerd/pull/2366, as I was hoping for some kind of upstream movement before I pushed too much further down a certain direction.

So what I pushed to my fork is the latest I have, which worked-around or ignored those issues, and there's plenty that can be done in areas which do not relate to them, like the networking setup, the filesystem-copy/permissions issue, etc.

I suspect getting it upstreamed will be equally as hard as actually getting it working, in the end.
